### ISSUE TITLE
Ignore type of `_` in compatibility checks during subclassing

### ISSUE THREAD

**Issue by Azureblade3808** [microsoft/pyright#3574]
## Environment data

-   Language Server version: v2022.6.11 Pre-Release
-   OS and version: Windows 7
-   Python version (& distribution if applicable, e.g. Anaconda): 3.8.11 Anaconda

## Code Snippet

```python
class Base0:
    _: int = 0
class Derived0(Base0):
    _: str = ""

class Base1:
    def _(self, x: int) -> None:
        ...
class Derived1(Base1):
    def _(self, x: str) -> None:
        ...

class Base2:
    _: int = 0
class Derived2(Base2):
    def _(self) -> None:
        ...

class Base3:
    def _(self) -> None:
        ...
class Derived3(Base3):
    _: int = 0
```

## Expected behavior

`_` members get special-treated and no errors/warnings get reported for the code above.

## Actual behavior

Every `_` in a derived class gets reported as incompatible with super.

**Comment by erictraut** [comment]
What is the justification for ignoring compatibility checks for `_`? I know that there is a convention in Python for using `_` to indicate "a symbol whose value I don't care about", but that doesn't appear to be how you're using it here.

The `reportIncompatibleMethodOverride` diagnostic check is a strict-mode rule, so I think it's reasonable for it to be somewhat opinionated. If you want to suppress it within a particular source file, you can use a `# pyright: reportIncompatibleMethodOverride=false` comment at the top of that file.

**Comment by Azureblade3808** [comment]
I don't think there is much difference between
```python
class Base:
    def _(self, x: int) -> None:
        ...
    def _(self, x: str) -> None:
        ...
class Derived(Base):
    pass
```
and
```python
class Base:
    def _(self, x: int) -> None:
        ...
class Derived(Base):
    def _(self, x: str) -> None:
        ...
```
, as long as these `_` members don't get read outside the scope of class definition.

If such `_` members are considered as legitmate class members, in order that code like `something._(...)` may work, I'm all for current checks.

But some logic still could be added about `del` statements, so that I may at least write code like
```python
class Base:
    def _(self, x: int) -> None:
        ...
    del _
class Derived(Base):
    def _(self, x: str) -> None:
        ...
    del _
```
to make it work.

**Comment by erictraut** [comment]
I still don't understand why you want to name methods `_` and override the `_` method of a base class with an incompatible method of the same name? What is the reason for doing this? Is this for single dispatch design patterns?

**Comment by Azureblade3808** [comment]
Here is roughly what I've done --
```python
from __future__ import annotations

from typing import Callable, Sequence, TypeVar

import rx, rx.subject

Any_t = TypeVar("Any_t")
Callable_t = TypeVar("Callable_t", bound=Callable)

def supports_rx_bindings(cls: type[Any_t], /) -> type[Any_t]:
    # Alter `__init__` on the given class, appending binding logic for the recorded
    # key-paths and callbacks.
    ...

def rx_binding(key_path: str | Sequence[str], /) -> Callable[[Callable_t], Callable_t]:
    # When the returned callable, which would be used as an annotation, gets called,
    # use its argument as a callback, then record the key-path and the callback into
    # an internal list.
    ...

@supports_rx_bindings
class C:
    rx_a: rx.typing.Subject[int, int]
    recorded_a_values: list[int]

    rx_b: rx.typing.Subject[str, str]
    recorded_b_values: list[str]

    def __init__(self) -> None:
        self.rx_a = rx.subject.BehaviorSubject(0)
        self.rx_b = rx.subject.BehaviorSubject("")

        self.recorded_a_values = []
        self.recorded_b_values = []

    @rx_binding(["a"])
    def _(self, a: int, /) -> None:
        self.recorded_a_values.append(a)

    @rx_binding(["b"])
    def _(self, b: str, /) -> None:
        self.recorded_b_values.append(b)

c = C()
assert c.recorded_a_values == [0]
assert c.recorded_b_values == [""]

c.rx_a.on_next(1)
assert c.recorded_a_values == [0, 1]

c.rx_b.on_next("abc")
assert c.recorded_b_values == ["", "abc"]

```

Naming for callbacks could be hard for observing a bundle of deep key-paths, and the callbacks usually don't get called manually aside of unit-tests. Therefore, the names don't matter in most cases. So I'd like to see `_` gets used as a common name for callbacks.

**Comment by erictraut** [comment]
This is included in pyright 1.1.255, which I just published. It will also be included in the next release of pylance.
